### ISSUE TITLE
Feature request: API to load local files

### ISSUE THREAD

**Issue by matthdsm** [igvteam/igv.js#568]
ref #311 

Would it be possible to add API functionality for:
* reading client side files
* moving buttons from IGV header bar to other UI elements.

Thanks a lot!
M

**Comment by el1jaC** [comment]
For the first point, do you want to load local files dynamically from "load" button, or loading local file on  UI loading? 
I would like to warn you that reading local file from browser is virtually a worst practice due to many security reasons.
It would be better to read blobs from URL by IGV.js API.

Fro the second point, do you want to drag buttons using the UI, or create buttons dynamically on UI loading?

**Comment by matthdsm** [comment]
Hi,

I'd like to be able to load local files to the application. This is already implemented as referenced in #311, but needs an API. I'm not very familiar with how everything should/can work, so the actual implementation is something I'd like to leave to more experienced minds.

As for the second question, I'd like to be able to remove the buttons currently in the IGV top bar, and create those buttons elsewhere on the page (e.g. outside of the IGV div), so I'd like to create buttons on page load. Removal could be done with a param in the global config, while the button functionality could be bound to a method like `igv.browser.toggleCursorGuide()` or something.

Thanks,
M

**Comment by el1jaC** [comment]
To load local files, API must upload them in a temporary folder (obviously in workspace), and import them in browser using **browser.loadTrack** linking  relative path.
I try to develop it. I keep you updated.

For the second point, buttons do you mean **_Track Labels, Center Line, Cursor Guide and zoom plus/minus_**?

**Comment by matthdsm** [comment]
Thanks a lot.
For the buttons, those are the ones indeed. The button aforementioned don't bother me all that much, it's mostly the "Load Track" button, as activated by enabling the `showLoadFileWidget` in the config.

Actually, if the functions bound to said widget had an API, I thinks all needed functionality should be there.

M

**Comment by jrobinso** [comment]
Hi Matthew,
Sorry for the delay in responding,  we've been busy.    Yes I agree every button (not just load file) should have an associated API.   We will do that.  I think you can already hide the nav bar, so 1/2 of it is done.    WRT loading local files,  we will support loading from a blob object,  same as the current load file widget produces.   This is the only legal way to load local files,  you cannot use file paths, or even know the file paths via javascript.   This you probably already know.

So in summary,  an API associated with every button,  including the "file" button,  along with an option to hide them all should satisfy 1 and 2.

**Comment by matthdsm** [comment]
Hi Jim,

Sounds awesome! Looking forward to testing the new functions? Will you be adding some docs to the wiki?

Thanks again
M

**Comment by el1jaC** [comment]
@jrobinso I can develop a simple "file" button with an API, just tell me when the button must be placed. I've already done it in JavaFX, and I don't think it will take much.

**Comment by jrobinso** [comment]
@el1jaC We already have solved this problem.  Mattias is requesting a load API.  

@matthdsm It occurs to me we might already have the API.   Try loadTrack, but put the file blob in the URL field.

**Comment by jrobinso** [comment]
Also,  @el1jaC is correct, you must get handles to both files for indexed formats and explicitly set indexURL (again with the file blob
).

**Comment by el1jaC** [comment]
@jrobinso so I think you should explain in the wiki how to link indexed file (with index file named equal to content file).

**Comment by turner** [comment]
Loading a local file is supported via the `loadTrack(config)` function. Here config is an object with a property named `url` that is set to a `File` object (which is a BLOB extension). The `File` object would typically be retrieved by navigating to the local file via - for example - the finder on a Mac.

**Comment by matthdsm** [comment]
Clear. Any pointers on how to get a File object from a form? I'm not all that experienced in JS and I'm having some issues getting it to work :/ 

Thank you for all the help
M

**Comment by turner** [comment]
There are tons of examples in the Web. Here is a sketch. 

I have a Bootstrap modal with an file load `input` element:
![screen shot 2018-03-29 at 9 09 33 am](https://user-images.githubusercontent.com/87169/38090852-d330e9dc-3331-11e8-8d12-82b57b135a57.png)

Code for the input element:
`<input id="my-local-file-widget" type="file" title="Load Local File"/>`

Click on the "Choose File" to present the finder (Mac example, similar for Windows):
![finder](https://user-images.githubusercontent.com/87169/38090927-0ce1bd6e-3332-11e8-823c-62f789b515c9.png)
